lighthouse-core/report/html/components/gauge.js

```diff
@@ -0,0 +1,398 @@
+  background-repeat: no-repeat;
+  background-size: var(--plugin-icon-size);
+  background-position: 58% 50%;
+  content: "";
+  position: absolute;
+  right: -6px;
+  bottom: 0px;
+  display: block;
+  z-index: 100;
+  box-shadow: 0 0 4px rgba(0,0,0,.2);
+  border-radius: 25%;
+}
+.lh-category .lh-gauge__wrapper--plugin .lh-gauge__svg-wrapper::before {
+  width: var(--plugin-badge-size-big);
+  height: var(--plugin-badge-size-big);
+}
+
+@keyframes load-gauge {
+  from { stroke-dasharray: 0 352; }
+}
+
+.lh-gauge__percentage {
+  width: 100%;
+  height: var(--gauge-circle-size);
+  position: absolute;
+  font-family: var(--report-font-family-monospace);
+  font-size: calc(var(--gauge-circle-size) * 0.34 + 1.3px);
+  line-height: 0;
+  text-align: center;
+  top: calc(var(--score-container-padding) + var(--gauge-circle-size) / 2);
+}
+
+.lh-category .lh-gauge__percentage {
+  --gauge-circle-size: var(--gauge-circle-size-big);
+  --gauge-percentage-font-size: var(--gauge-percentage-font-size-big);
+}
+
+.lh-gauge__wrapper {
+  position: relative;
+  display: flex;
+  align-items: center;
+  flex-direction: column;
+  text-decoration: none;
+  padding: var(--score-container-padding);
+
+  --circle-border-width: 8;
+  --transition-length: 1s;
+
+  /* Contain the layout style paint & layers during animation*/
+  contain: content;
+  will-change: opacity; /* Only using for layer promotion */
+}
+
+.lh-gauge__label {
+  font-size: var(--gauge-label-font-size);
+  line-height: var(--gauge-label-line-height);
+  margin-top: 10px;
+  text-align: center;
+  color: var(--report-text-color);
+}
+
+/* TODO(#8185) use more BEM (.lh-gauge__label--big) instead of relying on descendant selector */
+.lh-category .lh-gauge__label {
+  --gauge-label-font-size: var(--gauge-label-font-size-big);
+  --gauge-label-line-height: var(--gauge-label-line-height-big);
+  margin-top: 14px;
+}
+  </style>
+
+  <a href="#" class="lh-gauge__wrapper">
+    <!-- Wrapper exists for the ::before plugin icon. Cannot create pseudo-elements on svgs. -->
+    <div class="lh-gauge__svg-wrapper">
+      <svg viewBox="0 0 120 120" class="lh-gauge">
+        <circle class="lh-gauge-base" r="56" cx="60" cy="60"></circle>
+        <circle class="lh-gauge-arc" transform="rotate(-90 60 60)" r="56" cx="60" cy="60"></circle>
+      </svg>
+    </div>
+    <div class="lh-gauge__percentage"></div>
+    <div class="lh-gauge__label"></div>
+  </a>
+`;
+
+const attributes = {
+  score: '0.5',
+  variant: 'pass',
+  label: 'Lighthouse',
+};
+
+class LighthouseGaugeElement extends HTMLElement {
+  constructor() {
+    super();
+    this._shadowRoot = this.attachShadow({mode: 'open'});
+    this._shadowRoot.appendChild(template.content.cloneNode(true));
+    this._attributes = {...attributes};
+
+    for (const key in Object.keys(attributes)) {
+      const attributeName = /** @type {keyof typeof attributes} */ (key);
+      Object.defineProperty(this, attributeName, {
```

Comment:
Personally, I'm not sure the accessor generation is buying a lot here. First, it's a little wasteful to define them for every instance, you could just do it on the prototype once. Then if your'e doing that, I think that just writing the accessors in the class body is pretty readable, if you're going very vanilla components. That also makes the properties visible to static analysis tools:

```js
class LighthouseGaugeElement extends HTMLElement {

  static observedAttributes = ['score', 'variant', 'label'];

  get score() { return this.getAttribute('score'); }
  set score(value) { return this.setAttribute('score', value); }

  get variant() { return this.getAttribute('variant'); }
  set variant(value) { return this.setAttribute('variant', value); }

  get label() { return this.getAttribute('label'); }
  set label(value) { return this.setAttribute('label', value); }

  constructor() {
    super();
    this.score = '0.5';
    this.variant = 'pass';
    this.label = 'Lighthouse';
    // do template stuff...
  }

}
```

Reply:
yeah that's fair especially with only 3 here right now, is there a point at which you would recommend applying the pattern of dynamically generated properties ala lit-element?